src/main/java/com/aws/greengrass/certificatemanager/certificate/CISShadowMonitor.java

```diff
@@ -208,31 +252,173 @@ private void subscribeToShadowTopics() throws InterruptedException {
+        shadowProcessor.queueProcessingRequest(
+                ShadowProcessingRequest.builder()
+                        .version(response.version)
+                        .desiredState(response.state.desired)
+                        .build());
     }
 
-    @SuppressWarnings("PMD.AvoidCatchingGenericException")
-    private synchronized void processCISShadow(int version, Map<String, Object> desiredState) {
-        if (version == lastVersion) {
-            LOGGER.atInfo().kv(VERSION, version).log("Already processed version. Skipping cert re-generation");
-            updateCISShadowReportedState(version, desiredState);
-            return;
+    private void queueCISShadowProcessing(ShadowDeltaUpdatedEvent event) {
+        shadowProcessor.queueProcessingRequest(
+                ShadowProcessingRequest.builder()
+                        .version(event.version)
+                        .desiredState(event.state)
+                        .build());
+    }
+
+    @Data
+    @Builder
+    private static class ShadowProcessingRequest {
+        private final int version;
+        private final Map<String, Object> desiredState;
+    }
+
+    /**
+     * Responsible for processing CIS shadow state.
+     *
+     * <p>The goal of this class is to react to changes in CIS connectivity
+     * and rotate registered certificates using the new host information.
+     *
+     * <p>Shadow processing by this class is idempotent, in the sense that duplicate requests
+     * to process the same shadow version will not result in unnecessary certificate rotations.
+     * Certificate rotation will only occur if this class detects changes in connectivity info
+     * from CIS.
+     *
+     * <p>If certificate rotation is successful, CIS shadow reported state
+     * will be updated to match desired state.
+     */
+    @RequiredArgsConstructor
+    class ShadowProcessor {
+
+        /**
+         * Request from {@link CISShadowMonitor} to process shadow state.
+         */
+        private final AtomicReference<ShadowProcessingRequest> shadowProcessingRequest = new AtomicReference<>();
+
+        /**
+         * Last received CIS shadow version, used to de-duplicate processing requests.
+         */
+        private final AtomicInteger lastCISShadowVersion = new AtomicInteger();
+
+        /**
+         * Host addresses obtained from CIS, used to de-duplicate processing requests.
+         */
+        private final AtomicReference<List<String>> hostAddresses = new AtomicReference<>();
+
+        private final ScheduledExecutorService ses;
+
+        /**
+         * Scheduled task for handling shadow processing requests.
+         */
+        private Future<?> shadowProcessingTask;
+
+        /**
+         * Start processing shadow requests.
+         */
+        void start() {
+            if (shadowProcessingTask != null) {
+                shadowProcessingTask.cancel(true);
+            }
+            shadowProcessingTask = ses.scheduleWithFixedDelay(
```

Comment:
this check, cancel, and set needs to be synchronized. If this method can only ever be called by a single thread, then that's OK.

Reply:
it's only called as part of `ClientDevicesAuthService#startup`, but it wouldn't hurt to properly synchronize so we don't have to worry about it later